### PR TITLE
Add more project root inference for `versions` (aka `Unable to find project root package.json` issue)

### PR DESCRIPTION
**Pull request by ryan-roemer**
We have a bug opened in `inspectpack` repo that is really a `webpack-dashboard` issue. Fixes https://github.com/FormidableLabs/inspectpack/issues/45

The `inspectpack` `versions` action detects versions skews in installed packages within a webpack bundle by actually traversing the `node_modules` installed from the "project root" which is basically the directory that contains the operating `package.json`.

Some bespoke webpack configurations change `bundle.context` to something _other_ than a directory that is our "project root". This causes `inspectpack` to rightfully bomb out with a `Error: Unable to find project root package.json` error -- because if you're running `versions` you need to have this accessible.

In `webpack-dashboard`-land, it pays to be a little more permissive, because folks might not mind missing `versions` for getting all the _other_ cool dashboard stuff. So, accordingly, here's how the dashboard now deals with the problem:

1. If the `DashboardPlugin` specifies a `root` option to the root project directory, we use that. **Note**: We don't _check_ it's valid -- if it's invalid and manually specified you **will** get this same error.
2. If no `root` is specified, try to see if `bundle.context` has a `package.json`, use that.
3. If `bundle.context` isn't workable, try `process.cwd()`.
4. If `process.cwd()` isn't usable, then just _disable_ the `versions` action for the dashboard.

So, looking at @derkjn 's handy error repro repo https://github.com/derkjn/webpack-dashboard-error let's assume we are running webpack / commands from this directory:

https://github.com/derkjn/webpack-dashboard-error/tree/master/web/app/themes/ignitetheme

Then, we can edit https://github.com/derkjn/webpack-dashboard-error/blob/master/web/app/themes/ignitetheme/resources/assets/build/webpack.config.js with any of the following options and it will work without error:

Just like it exists now. Works because `process.cwd()` is used.

```js
  new DashboardPlugin()
```

Manually specify a root. These all basically resolve to `/web/app/themes/ignitetheme`.

```js
  new DashboardPlugin({
    root: process.cwd()
  })
```

```js
  new DashboardPlugin({
    // webpack config is three levels deeper than CWD
    root:  path.resolve(__dirname, "../../..")
  })
```

```js
  new DashboardPlugin({
    // This was set in a different file to the same thing.
    root: config.paths.root
  })
```

@derkjn -- Can you confirm this new version of `webpack-dashboard` now fixes your error?

/cc @kenwheeler @tptee

### CODE REVIEW

File: plugin/index.js [hunk 1]
```diff
@@ -34,101 +34,20 @@ function getTimeMessage(timer) {
   return ` (${time})`;
 }
 
-function observeBundleMetrics(stats, inspectpack) {
```
Comment (ryan-roemer) [inline plugin/index.js]: I need to get to the `root` plugin option, so this became a class method.

File: plugin/index.js [hunk 2]
```diff
@@ -34,101 +34,20 @@ function getTimeMessage(timer) {
...
       options = options || {};
       this.port = options.port || DEFAULT_PORT;
+      this.root = options.root;
       this.handler = options.handler || null;
     }
 
     this.cleanup = this.cleanup.bind(this);
 
     this.inspectpack = InspectpackDaemon.create({ cacheFilename });
-    console.log(this.inspectpack);
```
Comment (ryan-roemer) [inline plugin/index.js]: Tsk, tsk...

File: plugin/index.js [hunk 3]
```diff
@@ -263,16 +182,153 @@ class DashboardPlugin {
...
+      if (require(path.resolve("package.json"))) {
+        return process.cwd();
+      }
+    } catch (err) { /* passthrough */ }
+
+    // A null will be filtered out, disabling `versions` action.
+    return null;
+  }
+
+  observeBundleMetrics(stats, inspectpack) {
```
Comment (ryan-roemer) [inline plugin/index.js]: This is pretty much the original function, just now as a member method.

File: plugin/index.js [hunk 4]
```diff
@@ -263,16 +182,153 @@ class DashboardPlugin {
...
+          type: "sizes",
+          error: true,
+          value: serializeError(err)
+        }));
+
+    const getProblems = bundles =>
+      Promise.all(
+        INSPECTPACK_PROBLEM_ACTIONS.map(action =>
+          Promise.all(
+            bundles
```
Comment (ryan-roemer) [inline plugin/index.js]: I've actually changed stuff in L288-L313 -- testing project root and removing `versions` if there isn't a package.json

File: utils/format-problems.js [hunk 5]
```diff
@@ -16,7 +20,8 @@ function formatProblems(bundle) {
   if (!duplicates && versions) {
     return `${chalk.green("No duplicate files!")}\n${versions}`;
   }
-  return `${formatDuplicates(bundle.duplicates)}\n${formatVersions(bundle.versions)}`;
```
Comment (ryan-roemer) [inline utils/format-problems.js]: We were needlessly reformatting what we had already formatted in original L8-9.

**Comment by derkjn** [comment]
Hi @ryan-roemer , I just pulled down the `bug/versions-project-root` branch and tested it in my project and I can confirm it's all working perfectly now! Thank you for your help on this! 😄

**Comment by ryan-roemer** [comment]
Released in `webpack-dashboard@1.0.0-7`

**Pull request merged**